msm/src/prover.rs

```diff
@@ -66,45 +84,159 @@ where
     let mut fq_sponge = EFqSponge::new(G::other_curve_sponge_params());
 
     // Do not use parallelism
-    commitments
+    witness_comms
         .into_iter()
         .for_each(|comm| absorb_commitment(&mut fq_sponge, comm));
 
     // -- Start MVLookup
-    let lookup_env = if !inputs.mvlookups.is_empty() {
+    let lookup_env = if !witness.mvlookups.is_empty() {
         Some(mvlookup::prover::Env::create::<OpeningProof, EFqSponge>(
-            inputs.mvlookups,
+            witness.mvlookups,
             domain,
             &mut fq_sponge,
             srs,
         ))
     } else {
         None
     };
+
+    // Don't need to be absorbed. Already absorbed in mvlookup::prover::Env::create
+    // FIXME: remove clone
+    let mvlookup_comms = Option::map(lookup_env.as_ref(), |lookup_env| LookupProof {
+        m: lookup_env.lookup_counters_comm_d1.clone(),
+        h: lookup_env.lookup_terms_comms_d1.clone(),
+        sum: lookup_env.lookup_aggregation_comm_d1.clone(),
+    });
+
     // -- end computing the running sum in lookup_aggregation
     // -- End of MVLookup
 
-    // TODO: add quotient polynomial (based on constraints and expresion framework)
+    // TODO rename this
+    // The evaluations should be at least the degree of our expressions. Higher?
+    // Maybe we can only use d4, we don't have degree-7 gates anyway
+    let mut witness_evals_env: Vec<Evaluations<G::ScalarField, R2D<G::ScalarField>>> = vec![];
+    for witness_poly in witness_polys.x.iter() {
```

Comment:
You should be able to simple use `iter`. We want to abstract the Witness structure.

Reply:
See b8eb40e99c15572362213c3b3520034c0f62a54d.